cranelift/codegen/src/machinst/abi_impl.rs

```diff
@@ -0,0 +1,1146 @@
+
+    /// Generate a store to the stack.
+    fn gen_store_stack(mem: StackMemArg, from_reg: Reg, ty: Type) -> Self::I;
+
+    /// Generate a move.
+    fn gen_move(to_reg: Writable<Reg>, from_reg: Reg, ty: Type) -> Self::I;
+
+    /// Generate an integer-extend operation.
+    fn gen_extend(
+        to_reg: Writable<Reg>,
+        from_reg: Reg,
+        is_signed: bool,
+        from_bits: u8,
+        to_bits: u8,
+    ) -> Self::I;
+
+    /// Generate a return instruction.
+    fn gen_ret() -> Self::I;
+
+    /// Generate an "epilogue placeholder" instruction, recognized by lowering
+    /// when using the Baldrdash ABI.
+    fn gen_epilogue_placeholder() -> Self::I;
+
+    /// Generate an add-with-immediate. Note that even if this uses a scratch
+    /// register, the sequence must still be correct if the given source or dest
+    /// is the register returned by `get_fixed_tmp_reg()`; hence, for machines
+    /// that may need a scratch register to synthesize an arbitrary constant,
+    /// the machine backend should reserve *another* fixed temp register for
+    /// this purpose. (E.g., on AArch64, x16 is the ordinary fixed tmp, and x17
+    /// is the secondary fixed tmp used to implement this.)
+    fn gen_add_imm(into_reg: Writable<Reg>, from_reg: Reg, imm: u64) -> SmallVec<[Self::I; 4]>;
+
+    /// Generate a sequence that traps with a `TrapCode::StackOverflow` code if
+    /// the stack pointer is less than the given limit register (assuming the
+    /// stack grows downward).
+    fn gen_stack_lower_bound_trap(limit_reg: Reg) -> SmallVec<[Self::I; 2]>;
+
+    /// Generate an instruction to compute an address of a stack slot (FP- or
+    /// SP-based offset).
+    fn gen_get_stack_addr(mem: StackMemArg, into_reg: Writable<Reg>, ty: Type) -> Self::I;
+
+    /// Get a fixed (not used by regalloc) temp. This is needed for certain
+    /// sequences generated after the register allocator has already run.
+    fn get_fixed_tmp_reg() -> Reg;
+
+    /// Generate a store to the given [base+offset] address.
+    fn gen_load_base_offset(into_reg: Writable<Reg>, base: Reg, offset: i64, ty: Type) -> Self::I;
+
+    /// Generate a load from the given [base+offset] address.
+    fn gen_store_base_offset(base: Reg, offset: i64, from_reg: Reg, ty: Type) -> Self::I;
+
+    /// Adjust the stack pointer up or down.
+    fn gen_sp_reg_adjust(amount: i64) -> SmallVec<[Self::I; 2]>;
+
+    /// Generate a meta-instruction that adjusts the nominal SP offset.
+    fn gen_nominal_sp_adj(amount: i64) -> Self::I;
+
+    /// Generate the usual frame-setup sequence for this architecture: e.g.,
+    /// `push rbp / mov rbp, rsp` on x86-64, or `stp fp, lr, [sp, #-16]!` on
+    /// AArch64.
+    fn gen_prologue_frame_setup() -> SmallVec<[Self::I; 2]>;
+
+    /// Generate the usual frame-restore sequence for this architecture.
+    fn gen_epilogue_frame_restore() -> SmallVec<[Self::I; 2]>;
+
+    /// Generate a clobber-save sequence.
```

Comment:
Could you clarify a bit what `gen_clobber_save/restore` do?  I can kinda guess, but a bit more clarity here would be nice (maybe just an extra sentence).